### DISCUSSION TITLE
Generic models

### DISCUSSION THREAD

**Discussion by cmutel** [Depart-de-Sentier/sentier.dev#17] [General]
A generic model can operate on many different products. This is in contrast to normal transformational models, where the input and output terms (and therefore glossary URIs) are fixed.

Why is this a problem? The concept we have so far is that we *find* models based on the *terms* they claim they produce. This breaks down when the terms aren't fixed.

**Comment by cmutel** [comment]
It there an analogy here to Python decorators? Decorators are just fancy ways to apply a function to another function. So we could think of these models as being transitive function application, e.g.

```python
store(harvest(crop))
```

But who is responsible for finding when generic models need to be applied, and then applying them in the right way?

There are a lot of advantages to having a fixed graph structure, so instead of:

```
A -> generic -> B
```

We can call:

```
generic(A) -> B
```

**Comment by cmutel** [comment]
@schoi839 Items in the glossary can have links to other generic models which *could* be run. Then we could make the product with and without the generic model.

We could also attach the generic models to the top of a product hierarchy instead of doing it manually.

**Comment by cmutel** [comment]
~Examples of~ Specific generic models:

- Food processing
- Storage
  -  Freezing
  - Controlled atmposhere
- (Transport - special case considered out of scope for this discussion)
- Pollution control equipment
- Cooling (electricity production)
- Material processing
- Gasses
  - Compression
  - Decompression
- Losses
  - Leaks (pipelines)
  - Unmetered usage
  - Spoilage
  - Evaporation or other routine losses
- Packaging (B2B and B2C)
- Waste/wastewater generation and treatment
- Operations of products (and lifetimes)

**Reply by mklarmann** [reply]
Operations on maintenance of products

**Comment by cmutel** [comment]
One key would be whether the product is different in the eyes of the consumers. For example fish which is or has been frozen is considered to be different for some than fish which was only fresh. 

If the properties aren't changing, then we could only offer the annual average.

We are at a risk of making a system which is very complicated, so some simplicity here would be nice.
